[FlexAttention] Enable different qk and v head-dims

drisspg opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #134043

# Summary
Adds the option for the head dims to be different between QK and V tensors.

Fixes issue: https://github.com/pytorch/pytorch/issues/133674

V_DIM > QK_DIM is blocked by landing: https://github.com/triton-lang/triton/pull/4138 / https://github.com/triton-lang/triton/pull/4540

Into PyTorch's triton branch.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

drisspg committed:
[FlexAttention] Enable different qk and v head-dims

[ghstack-poisoned]

drisspg committed:
Update on "[FlexAttention] Enable different qk and v head-dims"


# Summary
Adds the option for the head dims to be different between QK and V tensors.

Local testing shows that when QK_HEAD_DIM  > V head dim this works great for the forward
Not when V > QK , still debugging

[ghstack-poisoned]

drisspg committed:
Update on "[FlexAttention] Enable different qk and v head-dims"


# Summary
Adds the option for the head dims to be different between QK and V tensors.

Local testing shows that when QK_HEAD_DIM  > V head dim this works great for the forward
Not when V > QK , still debugging

[ghstack-poisoned]

drisspg committed:
Update on "[FlexAttention] Enable different qk and v head-dims"


# Summary
Adds the option for the head dims to be different between QK and V tensors.

Local testing shows that when QK_HEAD_DIM  > V head dim this works great for the forward
Not when V > QK , still debugging

[ghstack-poisoned]

drisspg committed:
Update on "[FlexAttention] Enable different qk and v head-dims"


# Summary
Adds the option for the head dims to be different between QK and V tensors.

Fixes issue: https://github.com/pytorch/pytorch/issues/133674

[ghstack-poisoned]

drisspg committed:
Update on "[FlexAttention] Enable different qk and v head-dims"


# Summary
Adds the option for the head dims to be different between QK and V tensors.

Fixes issue: https://github.com/pytorch/pytorch/issues/133674

[ghstack-poisoned]

drisspg commented on torch/_inductor/kernel/flex_decoding.py:
```diff
@@ -136,21 +139,13 @@ def flex_decoding_grid(batch_size, kv_heads, gqa_group_size, n_keys, d_model, me
 
     q_range = stride_qg * off_g[:, None, None] + stride_qm * off_m[None, :, None] + stride_qk * offs_d[None, None, :]
 
-    Q_block_ptr = tl.make_block_ptr(
```
wasnt being used

drisspg committed:
Update on "[FlexAttention] Enable different qk and v head-dims"


# Summary
Adds the option for the head dims to be different between QK and V tensors.

Fixes issue: https://github.com/pytorch/pytorch/issues/133674

[ghstack-poisoned]

Chillee commented on torch/nn/attention/flex_attention.py:
```diff
@@ -848,6 +848,29 @@ def _apply_kernel_options(query, key, value, kernel_options):
+            f"Expect query and key/value to have the same embedding dimension "
+            f"but got E={query.size(-1)} and E={key.size(-1)}."
+        )
+    # TODO this config segfaults with Triton without:
```
I think you need to move this comment down.

drisspg committed:
Update on "[FlexAttention] Enable different qk and v head-dims"


# Summary
Adds the option for the head dims to be different between QK and V tensors.

Fixes issue: https://github.com/pytorch/pytorch/issues/133674

[ghstack-poisoned]

drisspg commented:
@pytorchbot merge

drisspg committed:
Update on "[FlexAttention] Enable different qk and v head-dims"


# Summary
Adds the option for the head dims to be different between QK and V tensors.

Fixes issue: https://github.com/pytorch/pytorch/issues/133674

V_DIM > QK_DIM is blocked by landing: https://github.com/triton-lang/triton/pull/4138 / https://github.com/triton-lang/triton/pull/4540

Into PyTorch's triton branch.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

drisspg committed:
Update

[ghstack-poisoned]

drisspg commented:
@pytorchbot merge

drisspg committed:
Update on "[FlexAttention] Enable different qk and v head-dims"


# Summary
Adds the option for the head dims to be different between QK and V tensors.

Fixes issue: https://github.com/pytorch/pytorch/issues/133674

V_DIM > QK_DIM is blocked by landing: https://github.com/triton-lang/triton/pull/4138 / https://github.com/triton-lang/triton/pull/4540

Into PyTorch's triton branch.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

drisspg committed:
Update on "[FlexAttention] Enable different qk and v head-dims"


# Summary
Adds the option for the head dims to be different between QK and V tensors.

Fixes issue: https://github.com/pytorch/pytorch/issues/133674

V_DIM > QK_DIM is blocked by landing: https://github.com/triton-lang/triton/pull/4138 / https://github.com/triton-lang/triton/pull/4540

Into PyTorch's triton branch.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

drisspg commented:
@pytorchbot merge

Skylion007 commented on torch/nn/attention/_utils.py:
```diff
@@ -31,6 +31,16 @@ def _calculate_scale(head_dim_size: int, scale: Optional[float]) -> float:
+    """SymInts dont support bitwise and, forgive me for my sins
+    These powers of 2 are realistically larger than what can fit in shared memory
+    """
+    return n in _POWERS_OF_2
```
Ugh, can't you just log base 2 to check if it's a power of 2?

drisspg replied:
yeah honestly , I am just going to rename this to "supported_head_dims" since its more accurate and doesnt feel as gross ahah

drisspg commented:
@pytorchbot merge

drisspg committed:
Update on "[FlexAttention] Enable different qk and v head-dims"


# Summary
Adds the option for the head dims to be different between QK and V tensors.

Fixes issue: https://github.com/pytorch/pytorch/issues/133674

V_DIM > QK_DIM is blocked by landing: https://github.com/triton-lang/triton/pull/4138 / https://github.com/triton-lang/triton/pull/4540

Into PyTorch's triton branch.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

drisspg commented:
@pytorchbot merge

drisspg committed:
Update on "[FlexAttention] Enable different qk and v head-dims"


# Summary
Adds the option for the head dims to be different between QK and V tensors.

Fixes issue: https://github.com/pytorch/pytorch/issues/133674

V_DIM > QK_DIM is blocked by landing: https://github.com/triton-lang/triton/pull/4138 / https://github.com/triton-lang/triton/pull/4540

Into PyTorch's triton branch.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

drisspg commented:
@pytorchbot merge

yanboliang commented:
@pytorchbot merge

yanboliang commented:
@pytorchbot merge -f "stucked ROCM jobs, flex attention unit tests only on CUDA"

yanboliang commented:
@pytorchbot merge -f "stucked ROCM jobs, flex attention unit tests only on CUDA"

yanboliang commented:
@pytorchbot merge
